bdk_core/tests/test_sparse_chain.rs

```diff
@@ -1,529 +1,361 @@
-        vec![TxSpec {
-            inputs: vec![ISpec::Other],
-            outputs: vec![OSpec::Mine(2_000, 0)],
-            confirmed_at: Some(0),
-        }],
-        0,
-    );
-
-    assert_eq!(chain.apply_checkpoint(update.clone()), ApplyResult::Ok);
-    update.base_tip = Some(update.new_tip);
-    update.new_tip = BlockId {
-        height: 1,
-        ..Default::default()
-    };
-    assert_eq!(chain.apply_checkpoint(update.clone()), ApplyResult::Ok);
-    assert_eq!(chain.iter_checkpoints(..).count(), 2);
-
-    update.base_tip = Some(update.new_tip);
-    update.new_tip = BlockId {
-        height: 2,
-        ..Default::default()
-    };
-    assert_eq!(chain.apply_checkpoint(update.clone()), ApplyResult::Ok);
-    assert_eq!(chain.iter_checkpoints(..).count(), 2);
-    assert_eq!(chain.iter_checkpoints(..).next().unwrap().height, 0);
-    assert_eq!(chain.iter_checkpoints(..).last().unwrap().height, 2);
+        chain.apply_checkpoint(CheckpointCandidate::new(last_valid, last_valid.unwrap())),
+        ApplyResult::Ok,
+        "repeated last_tip should succeed"
+    );
+
+    // ensure state of sparsechain is correct
+    chain
+        .iter_checkpoints(..)
+        .zip(0..)
+        .for_each(|(block_id, exp_height)| {
+            assert_eq!(block_id, gen_block_id(exp_height, exp_height as _))
+        });
 }
 
 #[test]
-fn adding_checkpoint_where_tx_conftime_has_changed() {
-    let mut checkpoint_gen = CheckpointGen::new();
+fn checkpoint_limit_is_respected() {
     let mut chain = SparseChain::default();
+    chain.set_checkpoint_limit(Some(5));
 
-    let mut update = checkpoint_gen.create_update(
-        vec![TxSpec {
-            inputs: vec![ISpec::Other],
-            outputs: vec![OSpec::Mine(1_900, 0)],
-            confirmed_at: Some(0),
-        }],
-        0,
-    );
+    // gen 10 checkpoints
+    let mut last_valid = None;
+    for i in 0..10 {
+        let new_tip = gen_block_id(i, i as _);
+        assert_eq!(
+            chain.apply_checkpoint(CheckpointCandidate {
+                txids: vec![(gen_hash(i as _), Some(i)),],
+                ..CheckpointCandidate::new(last_valid, new_tip)
+            }),
+            ApplyResult::Ok,
+        );
+        last_valid = Some(new_tip);
+    }
 
-    assert_eq!(chain.apply_checkpoint(update.clone()), ApplyResult::Ok);
-    update.base_tip = Some(update.new_tip);
-    update.new_tip = BlockId {
-        height: 1,
-        ..Default::default()
-    };
-    update.transactions[0].confirmation_time = Some(BlockTime { height: 1, time: 1 });
-    assert!(matches!(
-        chain.apply_checkpoint(update.clone()),
-        ApplyResult::Inconsistent { .. }
-    ));
-    assert_eq!(chain.iter_checkpoints(..).count(), 1);
+    assert_eq!(chain.iter_confirmed_txids().count(), 10);
+    let latest = chain.latest_checkpoint();
+    assert_eq!(chain.iter_checkpoints(..).count(), 5);
+    assert_eq!(chain.latest_checkpoint(), latest);
```

Comment:
I don't think you meant to check the variable you assigned two lines above.

Reply:
What was I thinking?